packages/fxa-react/lib/hooks.test.tsx

```diff
@@ -68,3 +74,202 @@ describe('useClickOutsideEffect', () => {
     expect(onDismiss).not.toBeCalled();
   });
 });
+
+describe('useAwait', () => {
+  const Subject = ({
+    fn = async () => 'test',
+    fnArgs,
+    initialState,
+    executeImmediately,
+    rethrowError,
+  }: {
+    fn?: (...args: any) => Promise<string>;
+    fnArgs?: any;
+    initialState?: PromiseState<string, any>;
+    executeImmediately?: boolean;
+    rethrowError?: boolean;
+  }) => {
+    const useDefaults = [initialState, executeImmediately, rethrowError].every(
+      (item) => typeof item === 'undefined'
+    );
+    const options = useDefaults
+      ? undefined
+      : {
+          initialState,
+          executeImmediately,
+          rethrowError,
+        };
+    const [state, execute, reset] = useAwait(fn, options);
+    const [thrownState, setThrownState] = useState('');
+    const executeMaybeThrows = useCallback(async () => {
+      try {
+        await execute(fnArgs);
+      } catch (e) {
+        setThrownState(e);
+      }
+    }, [execute]);
+    return (
+      <div>
+        {state.pending && <div data-testid="pending">pending</div>}
+        {!state.pending && <div data-testid="not-pending">not-pending</div>}
+        <div data-testid="state">{JSON.stringify(state)}</div>
+        <div data-testid="thrown">{JSON.stringify(thrownState)}</div>
+        <button data-testid="execute" onClick={executeMaybeThrows}></button>
+        <button data-testid="reset" onClick={reset}></button>
+      </div>
+    );
+  };
+
+  const parseState = (getByTestId: RenderResult['getByTestId']) =>
+    JSON.parse(getByTestId('state').textContent as string);
+
+  it('accepts initial state', async () => {
+    const initialState: PromiseStateResolved<string> = {
+      result: 'expected',
+      pending: false,
+      error: undefined,
+    };
+    const { getByTestId } = render(<Subject {...{ initialState }} />);
+    expect(parseState(getByTestId)).toEqual(initialState);
+  });
+
+  it('executes immediately when executeImmediately = true', async () => {
+    const expected = 'asdasdasd';
+    const { findByTestId, getByTestId } = render(
+      <Subject {...{ fn: async () => expected, executeImmediately: true }} />
+    );
+    await findByTestId('pending');
+    await findByTestId('not-pending');
```

Comment:
What do you think about using `screen` instead of destructuring these?

```
render(
  <Subject {...{ fn: async () => expected, executeImmediately: true }} />
);
await screen.findByTestId('pending');
await screen.findByTestId('not-pending');
```

I know Jody and I have preferred this syntax and started going this route with new tests but we didn't exactly bring it up as a potential new standard.

Reply:
Hmm I kind of don't like that syntax, if only for the redundant `screen` everywhere. If it made things more clear, maybe? But I think the function name is pretty verbose about it's purpose in the scope already.

I think it's also technically a `RenderResult` type, since the result of render can be an individual component and not a full screen of content. We use `renderResult` as a name [in a few spots](https://github.com/mozilla/fxa/search?q=renderResult&unscoped_q=renderResult) where that needs to be passed around as an intact structure

Reply:
That suggestion came from reading over this article: https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen

I've found it to be nice not having to update the destructure and to see a list of available methods on `screen`.  Though, hmm, I suppose you can see that while destructuring as well if you type one letter.

Reply:
Oh, that actually squicks out me a little bit - looks like he's using a `screen` module-global variable rather than `const screen = render(...)`. Missed that on first reading. Maybe a preference, but I *really* don't like that

Reply:
🤣 Maybe so. I know @jodyheavener likes it, I've liked it as well but as long as we're consistent I'm fine with either approach

Reply:
@lmorchard I'd be curious to know more about your dislike of it. If I understand correctly `screen` best reflects what a user actually sees, versus an isolated rendering. Is there a downside I'm not aware of?

Reply:
I just poked through the Jest documentation and all of their examples under "queries" use it. As it seems to be the way to query `document.body`, maybe it makes sense in some cases and less in others. Though, still might come down to personal preferences as in most test cases we render components in isolation. https://testing-library.com/docs/dom-testing-library/api-queries#screen